File: permute.py
def permute(a, i,n):

    if i==n:
        print(a)
    else:
        for j in range(i,n):
            a[i],a[j] = a[j],a[i]
            permute(a,i+1,n)
            a[i],a[j] = a[j],a[i]


permute(list(range(4)),0,4)

File: permute.c



void permute(char a[], int i, int n)
{
    int j;
    if (i == n)                  // If we've chosen all the characters then:
       cout << a << endl;        // we're done, so output it
    else
    {
        for (j = i; j <= n; j++) // Otherwise, we've chosen characters a[0] to a[j-1]
        {                        // so let's try all possible characters for a[j]
            swap(a[i], a[j]);    // Choose which one out of a[j] to a[n] you will choose
            permute(a, i+1, n);  // Choose the remaining letters
            swap(a[i], a[j]);    // Undo the previous swap so we can choose the next possibility for a[j]
        }
    }
}